pkg/finality-grandpa/lib.go

```diff
@@ -0,0 +1,398 @@
+// Copyright 2023 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"github.com/tidwall/btree"
+	"golang.org/x/exp/constraints"
+	"golang.org/x/exp/slices"
+)
+
+type HashNumber[Hash, Number any] struct {
+	Hash   Hash
+	Number Number
+}
+
+type targetHashTargetNumber[Hash, Number any] struct {
```

Comment:
Does this need to be different than HashNumber?

Reply:
They use these target attribute names in the rust implementation for `Prevote` and `Precommit`, and I think it makes sense to mimic the naming.